### PR TITLE
Create constructors as first method in class

### PR DESCRIPTION
**Pull request by shanethehat**
Currently when allowing phpspec to create a constructor, it places it as the last method in a class. This is because it uses the same logic as for placing a standard method, but by convention a constructor should be the first method in a class if one exists.

### CODE REVIEW

**Comment by ciaranmcnulty** [comment]
:+1: This has bugged me for ages.

**Comment by ciaranmcnulty** [comment]
Should probably do the same thing for named constructors, right?

**Comment by shanethehat** [comment]
Yeah, I probably should do that. What would be the convention there, private constructor first or named constructor method first?

**Comment by ciaranmcnulty** [comment]
I would say `__construct` first, then the others, then the normal methods

* New `__construct` goes at the top
* New private constructor goes at the top, but after `__construct` if it exists
* Other methods go at the end

File: src/PhpSpec/CodeGenerator/Generator/MethodGenerator.php [hunk 1]
```diff
@@ -23,6 +23,9 @@
  */
 class MethodGenerator implements GeneratorInterface
 {
+    const METHOD_PLACEMENT = '/}[ \n]*$/';
+    const CONSTRUCTOR_PLACEMENT = '/\n(?=\s*(private|protected|public)?\s?function)/';
```
Comment (ciaranmcnulty) [inline src/PhpSpec/CodeGenerator/Generator/MethodGenerator.php]: May contain `static`

File: src/PhpSpec/CodeGenerator/Generator/MethodGenerator.php [hunk 2]
```diff
@@ -111,4 +113,27 @@ protected function getTemplate()
...
+        return preg_replace(self::METHOD_PLACEMENT, rtrim($snippetToInsert)."\n}\n", trim($code));
+    }
+
+    /**
+     * @param $code
+     * @return bool
+     */
+    private function codeContainsAFunction($code)
+    {
+        return false !== strpos($code, 'function');
```
Comment (ciaranmcnulty) [inline src/PhpSpec/CodeGenerator/Generator/MethodGenerator.php]: This doesn't guarantee the code contains a function...

Comment (shanethehat) [inline src/PhpSpec/CodeGenerator/Generator/MethodGenerator.php]: it doesn't? Can you give an example of where it will fail?

Comment (ciaranmcnulty) [inline src/PhpSpec/CodeGenerator/Generator/MethodGenerator.php]: ^ edited in

File: src/PhpSpec/CodeGenerator/Generator/MethodGenerator.php (hunk 1 above)
Comment (shanethehat) [inline src/PhpSpec/CodeGenerator/Generator/MethodGenerator.php]: Or `final` ... anything else I've missed?

File: src/PhpSpec/CodeGenerator/Generator/MethodGenerator.php (hunk 2 above)
Comment (shanethehat) [inline src/PhpSpec/CodeGenerator/Generator/MethodGenerator.php]: Regex it is then I suppose. I was trying to avoid that.

Comment (ciaranmcnulty) [inline src/PhpSpec/CodeGenerator/Generator/MethodGenerator.php]: Regex still isn't that safe; we should probably use the tokenizer more for this stuff

File: features/code_generation/developer_generates_method.feature [hunk 3]
```diff
@@ -42,7 +42,6 @@ Feature: Developer generates a method
 
       class Markdown
       {
-
```
Comment (ciaranmcnulty) [inline features/code_generation/developer_generates_method.feature]: This makes me very happy

**Pull request merged**